internal/internal_workflow_client.go

```diff
@@ -1715,75 +1715,99 @@ func (w *workflowClientInterceptor) executeWorkflowWithOperation(
 			withStartOp,
 		},
 	}
-	multiResp, err := w.client.workflowService.ExecuteMultiOperation(ctx, &multiRequest)
 
-	var multiErr *serviceerror.MultiOperationExecution
-	if errors.As(err, &multiErr) {
-		if len(multiErr.OperationErrors()) != len(multiRequest.Operations) {
-			return nil, fmt.Errorf("%w: %v instead of %v operation errors",
-				errInvalidServerResponse, len(multiErr.OperationErrors()), len(multiRequest.Operations))
+	var startResp *workflowservice.StartWorkflowExecutionResponse
+	var updateResp *workflowservice.UpdateWorkflowExecutionResponse
+	for {
+		multiResp, err := func() (*workflowservice.ExecuteMultiOperationResponse, error) {
+			grpcCtx, cancel := newGRPCContext(ctx, grpcTimeout(pollUpdateTimeout), grpcLongPoll(true), defaultGrpcRetryParameters(ctx))
+			defer cancel()
+
+			multiResp, err := w.client.workflowService.ExecuteMultiOperation(grpcCtx, &multiRequest)
+			if err != nil {
+				if ctx.Err() != nil {
+					return nil, NewWorkflowUpdateServiceTimeoutOrCanceledError(err)
+				}
+				if status := serviceerror.ToStatus(err); status.Code() == codes.Canceled || status.Code() == codes.DeadlineExceeded {
+					return nil, NewWorkflowUpdateServiceTimeoutOrCanceledError(err)
+				}
+				return nil, err
+			}
+
+			return multiResp, err
+		}()
+
+		var multiErr *serviceerror.MultiOperationExecution
+		if errors.As(err, &multiErr) {
+			if len(multiErr.OperationErrors()) != len(multiRequest.Operations) {
+				return nil, fmt.Errorf("%w: %v instead of %v operation errors",
+					errInvalidServerResponse, len(multiErr.OperationErrors()), len(multiRequest.Operations))
+			}
+
+			var startErr error
+			var abortedErr *serviceerror.MultiOperationAborted
+			for i, opReq := range multiRequest.Operations {
+				// if an operation error is of type MultiOperationAborted, it means it was only aborted because
+				// of another operation's error and is therefore not interesting or helpful
+				opErr := multiErr.OperationErrors()[i]
+
+				switch t := opReq.Operation.(type) {
+				case *workflowservice.ExecuteMultiOperationRequest_Operation_StartWorkflow:
+					if !errors.As(opErr, &abortedErr) {
+						startErr = opErr
+					}
+				case *workflowservice.ExecuteMultiOperationRequest_Operation_UpdateWorkflow:
+					if !errors.As(opErr, &abortedErr) {
+						startErr = fmt.Errorf("%w: %w", errInvalidWorkflowOperation, opErr)
+					}
+				default:
+					// this would only happen if a case statement for a newly added operation is missing above
+					return nil, fmt.Errorf("%w: %T", errUnsupportedOperation, t)
+				}
+			}
+			return nil, startErr
```

Comment:
Can `startErr` be `nil` here? Even if server-side contract guarantees not, we might as well set a default somehow.

Reply:
Right, if the server misbehaves, it could be `nil`. I'll set a default 👍